src/utils/command.rs

```diff
@@ -41,75 +41,189 @@
+/// ```
+///
 /// ## Variant attributes
+/// All variant attributes override the corresponding `enum` attributes.
+///
 ///  1. `#[command(rename = "rule")]`
 /// Rename one command by a rule. Allowed rules are `lowercase`, `%some_name%`,
 /// where `%some_name%` is any string, a new name.
 ///
-///  2. `#[command(prefix = "prefix")]`
-/// Change a prefix for one command (the default is `/`).
+///  2. `#[command(parser = "parser")]`
+/// One more option is available for variants.
+///    - `custom_parser` - your own parser of the signature `fn(String) ->
+///    Result<Tuple, ParseError>`, where `Tuple` corresponds to the variant's
+/// arguments.
 ///
-///  3. `#[command(description = "description")]`
-/// Add a description of one command.
+/// ### Example
+/// ```
+/// use teloxide::utils::command::{BotCommand, ParseError};
+///
+/// fn accept_two_digits(input: String) -> Result<(u8), ParseError> {
+///     match input.len() {
+///         2 => {
+///             let num =
+///                 input.parse().map_err(|_| ParseError::IncorrectFormat)?;
+///             Ok((num))
+///         }
+///         len => Err(ParseError::Custom(format!(
+///             "Only 2 digits allowed, not {}",
+///             len
+///         ))),
+///     }
+/// }
+///
+/// #[derive(BotCommand, PartialEq, Debug)]
+/// #[command(rename = "lowercase", parser = "split")]
+/// enum Command {
+///     Num(u8),
+/// }
 ///
-/// All variant attributes overlap the `enum` attributes.
+/// let command = Command::parse("/num 12", "").unwrap();
+/// assert_eq!(command, Command::Num(12));
+/// let command = Command::parse("/num 333", "");
+/// assert!(command.is_err());
+/// ```
+///
+///  3. `#[command(prefix = "prefix")]`
+///  4. `#[command(description = "description")]`
+///  5. `#[command(separator = "sep")]`
+///
+/// Analogous to the descriptions above.
+///
+/// [`FromStr`]: https://doc.rust-lang.org/std/str/trait.FromStr.html
+/// [`BotCommand`]: crate::utils::command::BotCommand
 pub trait BotCommand: Sized {
-    fn try_from(s: &str) -> Option<Self>;
     fn descriptions() -> String;
-    fn parse<N>(s: &str, bot_name: N) -> Option<(Self, Vec<&str>)>
+    fn parse<N>(s: &str, bot_name: N) -> Result<Self, ParseError>
     where
         N: Into<String>;
 }
 
+/// Errors returned from [`BotCommand::parse`].
+///
+/// [`BotCommand::parse`]: crate::utils::command::BotCommand::parse
+#[derive(Debug)]
+pub enum ParseError {
+    TooFewArguments {
+        expected: usize,
+        found: usize,
+        message: String,
+    },
+    TooManyArguments {
+        expected: usize,
+        found: usize,
+        message: String,
+    },
+
+    /// Redirected from [`FromStr::from_str`].
+    ///
+    /// [`FromStr::from_str`]: https://doc.rust-lang.org/std/str/trait.FromStr.html#tymethod.from_str
+    IncorrectFormat,
+
+    UnknownCommand(String),
+    WrongBotName(String),
+
+    /// A custom error which you can return from your custom parser.
+    Custom(String),
```

Comment:
I think we'd better provide something like `Box<dyn Error>`, because it can be then downcasted if necessary. Nothing to do with just `String`.

Reply:
Ok